### PR TITLE
[type:feature] ShenYu Admin Cluster

### PR DESCRIPTION
**Pull request by Aias00**
See #5448 

Make sure that:

- [x] You have read the [contribution guidelines](https://shenyu.apache.org/community/contributor-guide).
- [x] You submit test cases (unit or integration tests) that back your changes.
- [x] Your local test passed `./mvnw clean install -Dmaven.javadoc.skip=true`.

### CODE REVIEW

**Comment by Aias00** [comment]
still working on

File: shenyu-admin/pom.xml [hunk 1]
```diff
@@ -84,6 +84,11 @@
             <artifactId>spring-integration-jdbc</artifactId>
         </dependency>
 
+        <dependency>
+            <groupId>org.springframework.integration</groupId>
+            <artifactId>spring-integration-zookeeper</artifactId>
```
Comment (moremind) [inline shenyu-admin/pom.xml:89]: this is should provided?

File: shenyu-admin/src/main/java/org/apache/shenyu/admin/config/ClusterConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+     * Shenyu running mode cluster service.
+     *
+     * @param shenyuClusterSelectMasterService shenyu cluster select master service
+     * @param upstreamCheckService upstream check service
+     * @param loadServiceDocEntry load service doc entry
+     * @param clusterProperties cluster properties
+     * @return Shenyu cluster service
+     */
+    @Bean(destroyMethod = "shutdown")
+    @ConditionalOnProperty(value = {"shenyu.cluster.enabled"}, havingValue = "true", matchIfMissing = true)
```
Comment (moremind) [inline shenyu-admin/src/main/java/org/apache/shenyu/admin/config/ClusterConfiguration.java:73]: may should same as `clusterForwardFilter`?

File: shenyu-admin/src/main/java/org/apache/shenyu/admin/config/ClusterJdbcConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+import javax.sql.DataSource;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * The type Cluster jdbc configuration.
+ */
+@Configuration(proxyBeanMethods = false)
+@EnableConfigurationProperties({ClusterProperties.class, ClusterZookeeperProperties.class})
+@ConditionalOnProperty(value = {"shenyu.cluster.select-type"}, havingValue = "jdbc", matchIfMissing = true)
```
Comment (moremind) [inline shenyu-admin/src/main/java/org/apache/shenyu/admin/config/ClusterJdbcConfiguration.java:46]: just use shenyu.cluster.type

**Review by moremind — CHANGES_REQUESTED** [review]

File: shenyu-admin/src/main/java/org/apache/shenyu/admin/listener/DataChangedEventDispatcher.java [hunk 4]
```diff
@@ -17,42 +17,67 @@
...
         this.applicationContext = applicationContext;
     }
 
     @Override
     @SuppressWarnings("unchecked")
-    public void onApplicationEvent(final DataChangedEvent event) {
+    public void onApplicationEvent(@NotNull final DataChangedEvent event) {
         for (DataChangedListener listener : listeners) {
+            if (!(listener instanceof AbstractDataChangedListener)) {
+                if (clusterProperties.isEnabled()
```
Comment (moremind) [inline shenyu-admin/src/main/java/org/apache/shenyu/admin/listener/DataChangedEventDispatcher.java:74]: change the condition xx&&yy&&zz

File: shenyu-admin/src/main/java/org/apache/shenyu/admin/listener/websocket/WebsocketCollector.java [hunk 5]
```diff
@@ -84,17 +88,50 @@ private static String getClientIp(final Session session) {
...
+            map.put("runningMode", runningMode);
+            map.put("masterUrl", masterUrl
+                    .replace("http", "ws")
+                    .replace("https", "ws")
+                    .concat("/websocket"));
+            if (isMaster) {
+                ThreadLocalUtils.put(SESSION_KEY, session);
+            }
 
+//            send(JsonUtils.toJson(map), DataEventTypeEnum.MYSELF);
```
Comment (moremind) [inline shenyu-admin/src/main/java/org/apache/shenyu/admin/listener/websocket/WebsocketCollector.java:121]: remove this

File: shenyu-admin/src/main/java/org/apache/shenyu/admin/mode/cluster/impl/jdbc/ClusterSelectMasterServiceJdbcImpl.java [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+        ClusterMasterDO masterDO = clusterMasterMapper.selectById(MASTER_ID);
+        return Objects.isNull(masterDO) ? new ClusterMasterDTO() : ClusterMasterTransfer.INSTANCE.mapToDTO(masterDO);
+    }
+    
+    @Override
+    public String getMasterUrl() {
+        ClusterMasterDO master = clusterMasterMapper.selectById(MASTER_ID);
+        String contextPath = master.getContextPath();
+        if (StringUtils.isEmpty(contextPath)) {
+            return "http://" + master.getMasterHost() + ":" + master.getMasterPort();
```
Comment (moremind) [inline shenyu-admin/src/main/java/org/apache/shenyu/admin/mode/cluster/impl/jdbc/ClusterSelectMasterServiceJdbcImpl.java:128]: consider the http schema

File: shenyu-admin/src/main/java/org/apache/shenyu/admin/mode/cluster/impl/zookeeper/ClusterSelectMasterServiceZookeeperImpl.java [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+import java.util.Map;
+import java.util.concurrent.locks.Lock;
+
+public class ClusterSelectMasterServiceZookeeperImpl implements ClusterSelectMasterService {
+    
+    private static final Logger LOG = LoggerFactory.getLogger(ClusterSelectMasterServiceZookeeperImpl.class);
+    
+    private static final String MASTER_LOCK_KEY = "cluster/master";
+    
+    private static final String MASTER_INFO = "/cluster/master/info";
```
Comment (moremind) [inline shenyu-admin/src/main/java/org/apache/shenyu/admin/mode/cluster/impl/zookeeper/ClusterSelectMasterServiceZookeeperImpl.java:40]: LOCK=shenyu_cluster_lock
MASTER_INFO = shenyu_cluster_lock/master/info
MASTER_LOCK_KEY = shenyu_cluster_lock/master

File: shenyu-admin/src/main/java/org/apache/shenyu/admin/mode/cluster/service/ShenyuClusterService.java [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+                TimeUnit.SECONDS);
+    }
+    
+    private void doSelectMaster(final String host, final String port, final String contextPath) {
+        // try getting lock
+        try {
+            boolean selected = shenyuClusterSelectMasterService.selectMaster(host, port, contextPath);
+            if (!selected) {
+                LOG.info("select master fail, wait for next period");
+//                TimeUnit.SECONDS.sleep(clusterProperties.getSelectPeriod());
```
Comment (moremind) [inline shenyu-admin/src/main/java/org/apache/shenyu/admin/mode/cluster/service/ShenyuClusterService.java:81]: remove this

File: shenyu-admin/src/main/java/org/apache/shenyu/admin/mode/cluster/service/ShenyuClusterService.java [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+//                try {
+                    // sleeps 10s then renew the lock
+                TimeUnit.SECONDS.sleep(clusterProperties.getSelectPeriod());
+                
+                renewed = shenyuClusterSelectMasterService.checkMasterStatus();
+                if (renewed) {
+                    LOG.info("renew master success");
+                }
+//                } catch (Exception e) {
+//                    // if renew fail, remove local master flag
```
Comment (moremind) [inline shenyu-admin/src/main/java/org/apache/shenyu/admin/mode/cluster/service/ShenyuClusterService.java:106]: remove this

File: shenyu-admin/src/main/java/org/apache/shenyu/admin/service/impl/UpstreamCheckService.java [hunk 10]
```diff
@@ -147,22 +147,22 @@ public UpstreamCheckService(final SelectorMapper selectorMapper,
...
      */
     public void setup() {
-        if (checked) {
-            this.fetchUpstreamData();
-            executor = new ScheduledThreadPoolExecutor(1, ShenyuThreadFactory.create("scheduled-upstream-task", false));
-            scheduledFuture = executor.scheduleWithFixedDelay(this::scheduled, 10, scheduledTime, TimeUnit.SECONDS);
-
-            ThreadFactory requestFactory = ShenyuThreadFactory.create("upstream-health-check-request", true);
-            invokeExecutor = new ScheduledThreadPoolExecutor(this.scheduledThreads, requestFactory);
+        if (REGISTER_TYPE_HTTP.equalsIgnoreCase(registerType)) {
```
Comment (moremind) [inline shenyu-admin/src/main/java/org/apache/shenyu/admin/service/impl/UpstreamCheckService.java:156]: change the conditon `if (REGISTER_TYPE_HTTP.equalsIgnoreCase(registerType) && checked)) {`

File: shenyu-sync-data-center/shenyu-sync-data-websocket/src/main/java/org/apache/shenyu/plugin/sync/data/websocket/client/ShenyuWebsocketClient.java [hunk 11]
```diff
@@ -129,71 +137,118 @@ public boolean connectBlocking() {
...
             send(DataEventTypeEnum.MYSELF.name());
             alreadySync = true;
         }
     }
-
+    
     @Override
     public void onMessage(final String result) {
-        handleResult(result);
+//        LOG.info("onMessage server[{}] result({})", this.getURI().toString(), result);
```
Comment (moremind) [inline shenyu-sync-data-center/shenyu-sync-data-websocket/src/main/java/org/apache/shenyu/plugin/sync/data/websocket/client/ShenyuWebsocketClient.java:153]: open this?

File: shenyu-sync-data-center/shenyu-sync-data-websocket/src/main/java/org/apache/shenyu/plugin/sync/data/websocket/client/ShenyuWebsocketClient.java [hunk 12]
```diff
@@ -129,71 +137,118 @@ public boolean connectBlocking() {
...
+        Map<String, Object> jsonToMap = JsonUtils.jsonToMap(result);
+        Object eventType = jsonToMap.get("eventType");
+        if (Objects.equals("CLUSTER", eventType)) {
```
Comment (moremind) [inline shenyu-sync-data-center/shenyu-sync-data-websocket/src/main/java/org/apache/shenyu/plugin/sync/data/websocket/client/ShenyuWebsocketClient.java:156]: not use magic string, you can create constant variable or enum include `CLUSTER`, `STANDALONE`

**Review by moremind — CHANGES_REQUESTED** [review]

**Review by moremind — APPROVED** [review]

**Pull request merged**